extension/agenthealth/handler/stats/provider/statuscode.go

```diff
@@ -0,0 +1,216 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package provider
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/amazon-contributing/opentelemetry-collector-contrib/extension/awsmiddleware"
+
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/agent"
+)
+
+const (
+	statusResetInterval = 5 * time.Minute
+	statusHandlerID     = "cloudwatchagent.StatusCodeHandler"
+	channelBufferSize   = 100
+)
+
+var (
+	statusCodeProviderSingleton *StatsProvider
+	statsProviderOnce           sync.Once
+)
+
+// StatsProvider handles processing of status codes and maintains stats.
+type StatsProvider struct {
+	statsByOperation map[string]*[5]int
+	resetTimer       *time.Timer
+	filter           agent.OperationsFilter
+	statusCodeChan   chan statusCodeEntry
+	stopChan         chan struct{}
+	mu               sync.Mutex
+	wg               sync.WaitGroup
+}
+
+// statusCodeEntry represents a status code and its associated operation.
+type statusCodeEntry struct {
+	operation  string
+	statusCode int
+}
+
+// GetStatusCodeStatsProvider initializes and retrieves the singleton StatsProvider.
+func GetStatusCodeStatsProvider(filter interface{}) *StatsProvider {
+	statsProviderOnce.Do(func() {
+		provider := &StatsProvider{
+			statsByOperation: make(map[string]*[5]int),
+			statusCodeChan:   make(chan statusCodeEntry, channelBufferSize),
+			stopChan:         make(chan struct{}),
+		}
+
+		if opsFilter, ok := filter.(agent.OperationsFilter); ok {
+			provider.filter = opsFilter
+		}
+		provider.startResetTimer()
+		provider.startProcessing()
+		statusCodeProviderSingleton = provider
+	})
+	return statusCodeProviderSingleton
+}
+
+// startProcessing begins processing status codes from the channel.
+func (sp *StatsProvider) startProcessing() {
+	sp.wg.Add(1)
+	go func() {
+		defer sp.wg.Done()
+		for {
+			select {
+			case entry := <-sp.statusCodeChan:
+				sp.processStatusCode(entry)
+			case <-sp.stopChan:
+				return
+			}
+		}
+	}()
+}
+
+// Stop signals the StatsProvider to stop processing and waits for cleanup.
+func (sp *StatsProvider) Stop() {
```

Comment:
This is never called and if it were we would need to ensure that it isn't called multiple times.